src/components/ProgressIndicator/ProgressIndicator.jsx

```diff
@@ -1,141 +1,364 @@
+  mainStep: PropTypes.bool,
+  subStep: PropTypes.bool,
+};
+
+ProgressStep.defaultProps = {
+  label: null,
+  secondaryLabel: null,
+  description: null,
+  onChange: null,
+  disabled: false,
+  showLabel: false,
+  stepWidth: null,
+  vertical: false,
+  invalid: false,
+  stepNumber: 0,
+  lastItem: false,
+  isClickable: false,
+  current: false,
+  complete: false,
+  incomplete: false,
+  mainStep: false,
+  subStep: false,
+};
+
 const ProgressIndicator = ({
+  className,
   items,
-  showLabels,
   currentItemId,
-  onClickItem,
-  stepWidth,
-  className,
+  showLabels,
   isVerticalMode,
+  stepWidth,
+  isClickable,
+  setStep,
+  onClickItem,
 }) => {
-  const handleChange = index => {
-    if (onClickItem) {
-      // Parent components are expecting the id not the index
-      onClickItem(items[index].id);
+  const [currentStep, setCurrentStep] = useState(currentItemId || items[0].id);
+
+  const flattenItems = (itemsList, level = 0) => {
+    let newList = [];
+    let step = 1;
+    let newVal;
+
+    itemsList.forEach(item => {
+      newVal = Object.assign({}, item);
+      if (item.children) {
+        delete newVal.children;
+        newVal.stepNumber = step;
+        newVal.level = level;
+        newList.push(newVal);
+        step += 1;
+        newList = newList.concat(flattenItems(item.children, level + 1));
+      } else {
+        newVal.stepNumber = step;
+        newVal.level = level;
+        newList.push(newVal);
+        step += 1;
+      }
+    });
+    return newList;
+  };
+
+  const newItems = items && flattenItems(items);
+
+  const lastItemId = () => newItems[newItems.length - 1].id;
+
+  const getCurrentIndex = () => {
+    let idx = 0;
+    if (currentStep) {
+      const index = newItems.findIndex(item => item.id === currentStep);
+      idx = index > -1 ? index : 0;
+    }
+    return idx;
+  };
+
+  const handleChange = step => {
+    if (step !== currentStep) {
+      if (setStep) {
+        setStep(step);
+      } else {
+        setCurrentStep(step);
+      }
     }
   };
 
-  const matchingIndex = useMemo(() => items.findIndex(item => item.id === currentItemId), [
-    items,
-    currentItemId,
-  ]);
+  useEffect(() => setCurrentStep(currentItemId), [currentItemId]);
 
-  // Only recalculate current step if inputs change
-  const currentStep = matchingIndex > -1 ? matchingIndex : 0;
-  return (
-    <StyledProgressIndicator
-      className={classnames(className, `${iotPrefix}--progress-indicator`)}
-      onChange={handleChange}
-      currentIndex={currentStep}
-      isVerticalMode={isVerticalMode}
-      vertical={isVerticalMode}
+  const classes = classnames({
+    [`${iotPrefix}--progress-indicator`]: true,
+    [`${iotPrefix}--progress-indicator--vertical`]: isVerticalMode,
+    [className]: className,
+  });
+
+  return newItems.length > 1 ? (
+    <ul
+      className={classes}
+      data-testid={`${iotPrefix}--progress-indicator-testid`}
+      onChange={onClickItem || handleChange}
```

Comment:
To preserve the original signature of this we need to call `onClickItem` with the `onClickItem(items[index].id);`, correct?

Reply:
No, by default the progress step onclick event will return the id instead of the index. Before we had a function called handleChange that would use items[index].id and pass it on to the callback function. But I have skipped that part as I am directly passing the id from the progress step.